modules/affile/directory-monitor-inotify.c

```diff
@@ -106,6 +106,11 @@ directory_monitor_inotify_new(const gchar *dir, guint recheck_time)
   self->super.stop_watches = _stop_watches;
   self->super.free_fn = _free;
   IV_INOTIFY_INIT(&self->inotify);
-  iv_inotify_register(&self->inotify);
+  if (iv_inotify_register(&self->inotify))
+    {
+      msg_error("directory-monitor-inotify: could not create inotify object", evt_tag_errno("errno", errno));
+      g_free(self);
```

Comment:
In the `directory_monitor_init_instance` the `dir` is `strdup`-ed, so in this case we leak that memory.
I would advice to call the free method, but the `_free` actually also do not free the `dir`, and also other possible members because parent free is not called.

Reply:
This surprised me as well: the reason why the parent free is not called from _free of DirectoryMonitorInotify is: she passes her free method to her parent as a virtual function. For me this is confusing: this latter suggests a plugin style relationship between DirectoryMonitor and DirectoryMonitorInotify, though the implementation of directory_monitor_inotify_new and struct DirectoryMonitorInotify suggests inheritance kind of relationship. This should be cleared first, because otherwise we cannot decide whose responsibility to free the DirectoryMonitor if the constructor of DirectoryMonitorInotify fails.

Probably we need to get rid of free_fn, and create proper init/deinit/new/free functions.

Reply:
In the end I overcomplicated the situation. There is directory_monitor_free() which could be called instead of g_free(). Until that part of the constructor, only the parent class is initialized, so we can simply call the parent's destructor. I updated the patchset.